src/mlio/image_reader.cxx

```
@@ -0,0 +1,392 @@
+    if (tsr == nullptr) {
+        return nullptr;
+    }
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(std::move(tsr));
+    return make_intrusive<example>(schema_, std::move(tensors));
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &instances,
+                          std::size_t batch_size) const
+{
+    std::atomic_bool skip_batch{};
+    auto instance_offset = params_.image_dimensions[0] *
+                           params_.image_dimensions[1] *
+                           params_.image_dimensions[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
+    std::vector<std::uint8_t> img_vec(total_memory_buffer);
+    auto *ptr_img_vec = img_vec.data();
+    int type;
+    cv::ImreadModes mode;
+    for (instance const &ins : instances) {
+        memory_slice img_buf;
+        if (params_.img_frame == image_frame::recordio) {
+            img_buf = ins.bits().subslice(recordio_image_header_offset_);
+        }
+        else {
+            img_buf = ins.bits();
+        }
+        cv::Mat mat(
+            1,
+            static_cast<int>(img_buf.size()),
+            CV_8U,
+            static_cast<void *>(const_cast<std::byte *>(img_buf.data())));
+        switch (params_.image_dimensions[0]) {
+        case 1:
+            mode = cv::ImreadModes::IMREAD_GRAYSCALE;
+            type = CV_8UC1;
+            break;
+        case 3:
+            mode = cv::ImreadModes::IMREAD_COLOR;
+            type = CV_8UC3;
+            break;
+        case 4:
+            mode = cv::ImreadModes::IMREAD_UNCHANGED;
+            type = CV_8UC4;
+            break;
+        default:
+            throw std::invalid_argument{
+                fmt::format("Unsupported number of channels: {0:n}",
+                            params_.image_dimensions[0])};
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins, skip_batch);
+        if (skip_batch) {
```

Instead of passing and checking `skip_batch`, can't we check whether `tmp` is `empty` and consider it as a failed decoding?